tsdb/head.go

```diff
@@ -652,6 +669,28 @@ Outer:
 			}
 			//nolint:staticcheck // Ignore SA6002 relax staticcheck verification.
 			tstonesPool.Put(v)
+		case []record.RefExemplar:
+			// todo (Callum): do we expect exemplar records to reach anywhere near the same level as samples,
+			// should they be processed in batches the same way samples are?
```

Comment:
Before we think about batches, I would suggest processing them in a separate goroutine like samples, but without batches. Since the exemplar storage is not totally optimised yet, this could become a bottleneck in the WAL replay.

Also you can add the caching of series like in processing samples to avoid multiple `getByID` calls.

Reply:
👍🏽

Reply:
I've made the change so that processing exemplars is done via a separate go routine, but adding the caching of the series actually didn't help performance at all for the `LoadWAL` benchmark, it actually just added allocs and bytes.